core/src/main/java/com/nisovin/magicspells/spells/passive/PotionEffectListener.java

```diff
@@ -0,0 +1,96 @@
+                    if (PotionEffectType.getByName(s) != null) { types.add(PotionEffectType.getByName(s)); }
+                    else MagicSpells.error("PotionEffect Passive " + spell.getInternalName() + " has an invalid effect defined: " + s + "!");
+                }
+            } else types = Arrays.asList(PotionEffectType.values()); //It's dirty, but it works. If a wildcard is used, dump every value into the list.
+
+            if (splits.length > 1 && !splits[1].equals("*")) {
+                for (String s : splits[1].split(",")) {
+                    try { actions.add(Action.valueOf(s)); }
+                    catch (IllegalArgumentException e) {MagicSpells.error("PotionEffect Passive " + spell.getInternalName() + " has an invalid action defined: " + s + "!");}
+                }
+            } else actions = Arrays.asList(Action.values());
+
+            if (splits.length > 1 && !splits[2].equals("*")) {
+                for (String s : splits[2].split(",")) {
+                    try { causes.add(Cause.valueOf(s)); }
+                    catch (IllegalArgumentException e) {MagicSpells.error("PotionEffect Passive " + spell.getInternalName() + " has an invalid cause defined: " + s + "!");}
+                }
+            } else causes = Arrays.asList(Cause.values());
+        }
+        spells.add(new potionTrigger(spell, types, actions, causes));
+    }
+
+    @EventHandler
+    public void onPotionEffect(EntityPotionEffectEvent event) {
+        if (!(event.getEntity() instanceof Player)) return;
+        Player player = (Player) event.getEntity();
+        Spellbook spellbook = MagicSpells.getSpellbook(player);
+        for (PotionEffectListener.potionTrigger trigger : spells) {
+            if (!isCancelStateOk(trigger.spell, event.isCancelled())) continue;
+            if (!spellbook.hasSpell(trigger.spell)) continue;
+            if (!trigger.actions.contains(event.getAction()) || !trigger.causes.contains(event.getCause())) continue;
+            PotionEffectType thisEffect = null;
+            switch (event.getAction()) { //The effect used by the event is referenced differently based on the action, so unfortunately this is needed
+                case ADDED:
+                    thisEffect = event.getNewEffect().getType();
+                    break;
+                case CHANGED:
+                    thisEffect = event.getModifiedType();
+                    break;
+                case REMOVED:
+                case CLEARED:
+                    thisEffect = event.getOldEffect().getType();
+                    break;
+            }
+            if (thisEffect.equals(null) || !trigger.types.contains(thisEffect)) continue;
+            boolean casted = trigger.spell.activate(player);
+            if (!PassiveListener.cancelDefaultAction(trigger.spell, casted)) continue;
+            event.setCancelled(true);
+        }
+    }
+
+    private static class potionTrigger {
```

Comment:
_Class names should be nouns, in mixed case with the first letter of each internal word capitalized_

potionTrigger should class be called PotionTrigger.